### PR TITLE
Adt/netcodev2

### PR DESCRIPTION
**Pull request by Vagabond**

### CODE REVIEW

File: src/game/game_state.c [hunk 1]
```diff
@@ -892,6 +892,42 @@ int game_state_ms_per_dyntick(game_state *gs) {
     return MS_PER_OMF_TICK;
 }
 
+int render_obj_clone(render_obj *src, render_obj *dst) {
+    memcpy(dst, src, sizeof(render_obj));
+    omf_calloc(1, sizeof(object));
```
Comment (mrannanj) [inline src/game/game_state.c:938]: this must be incorrect? return value of omf_calloc() is unused

File: src/game/protos/object.c [hunk 2]
```diff
@@ -146,6 +148,19 @@ int object_serialize(object *obj, serial *ser) {
...
 }
 
+int object_clone(object *src, object *dst) {
+    memcpy(dst, src, sizeof(object));
+    src->clone(src, dst);
+    if (src->custom_str) {
+        dst->custom_str = strdup(src->custom_str);
+    }
+
+    if (src->clone) {
```
Comment (mrannanj) [inline src/game/protos/object.c]: so here is a NULL-check for src->clone, but there the function is called before that, possibly crashing?

File: src/game/game_player.c [hunk 3]
```diff
@@ -26,11 +26,15 @@ void game_player_free(game_player *gp) {
 }
 
 void game_player_set_har(game_player *gp, object *har) {
-    gp->har = har;
+    if (har) {
```
Comment (mrannanj) [inline src/game/game_player.c:29]: Looks strange that this supports a NULL har, and is a void function, so caller has no idea what happened here..?

File: src/controller/net_controller.c [hunk 4]
```diff
@@ -106,6 +114,107 @@ void insert_event(list *transcript, int tick, uint16_t action, int id) {
...
+            // Handle waiting period leftover time
+            dynamic_wait -= 1;
+        }
+
+        // feed in the inputs
+        // XXX this is a hack for now
+
+        for (int j = 0; j < 2; j++) {
+            if (ev->events[j]) {
+                int player_id = abs(j);
```
Comment (mrannanj) [inline src/controller/net_controller.c]: `j` is either 0 or 1, no need for `abs()`

File: src/game/objects/hazard.c [hunk 5]
```diff
@@ -47,25 +46,27 @@ void hazard_tick(object *obj) {
...
+        void *userdata) {
+    scene *sc = (scene*)userdata;
 
     // Get next animation
-    bk_info *info = bk_get_info(&sc->bk_data, id);
+    bk_info *info = bk_get_info(sc->bk_data, id);
     if(info != NULL) {
-        object *obj = omf_calloc(1, sizeof(object));
-        object_create(obj, parent->gs, vec2i_add(pos, info->ani.start_pos), vel);
+        object *obj = malloc(sizeof(object));
```
Comment (mrannanj) [inline src/game/objects/hazard.c:56]: Why this change from omf_calloc to malloc?

File: src/game/protos/object.c [hunk 6]
```diff
@@ -554,20 +554,17 @@ void object_free(object *obj) {
...
-    if(obj->cur_animation_own == OWNER_OBJECT) {
-        animation_free(obj->cur_animation);
-        omf_free(obj->cur_animation);
-    }
     if(obj->custom_str) {
         omf_free(obj->custom_str);
     }
     obj->cur_surface = NULL;
     obj->cur_animation = NULL;
+    return 0;
```
Comment (mrannanj) [inline src/game/protos/object.c:424]: What's the point of int return value if it's always the same?

File: src/game/protos/scene.c [hunk 7]
```diff
@@ -154,7 +155,18 @@ int scene_unserialize(scene *s, serial *ser) {
...
+    }
+    return 0;
+}
+
+int scene_clone_free(scene *sc) {
+    if (sc->clone_free) {
+        sc->clone_free(sc);
+    }
+    memset(sc, 0, sizeof(scene));
+    omf_free(sc);
```
Comment (mrannanj) [inline src/game/protos/scene.c:145]: Why do we first memset something and then free it?

File: src/game/scenes/arena.c [hunk 8]
```diff
@@ -340,28 +340,29 @@ void arena_reset(scene *sc) {
...
-    object_set_stl(number, sc->bk_data.sound_translation_table);
+    object_set_stl(number, sc->bk_data->sound_translation_table);
     object_set_animation(number, number_ani);
     object_select_sprite(number, local->round);
     object_set_sprite_override(number, 1);
     game_state_add_object(sc->gs, number, RENDER_LAYER_TOP, 0, 0);
 }
 
 void arena_maybe_sync(scene *scene, int need_sync) {
+    return;
```
Comment (mrannanj) [inline src/game/scenes/arena.c]: Yeah, guess this is changing ..

File: src/game/utils/har_screencap.c [hunk 9]
```diff
@@ -20,6 +20,15 @@ void har_screencaps_reset(har_screencaps *caps) {
     har_screencaps_free(caps);
 }
 
+int har_screencaps_clone(har_screencaps *src, har_screencaps *dst) {
```
Comment (mrannanj) [inline src/game/utils/har_screencap.c:23]: void function

File: src/game/utils/score.c [hunk 10]
```diff
@@ -270,51 +270,17 @@ int chr_score_end_combo(chr_score *score, vec2i pos) {
     return ret;
 }
 
-void chr_score_serialize(chr_score *score, serial *ser) {
-    serial_write_int32(ser, score->score);
-    serial_write_int32(ser, score->done);
-    serial_write_int32(ser, score->scrap);
-    serial_write_int32(ser, score->destruction);
-    serial_write_int8(ser, score->texts.size);
+int chr_score_clone(chr_score *src, chr_score *dst) {
```
Comment (mrannanj) [inline src/game/utils/score.c:273]: void function

File: src/resources/animation.c [hunk 11]
```diff
@@ -53,6 +53,37 @@ animation *create_animation_from_single(sprite *sp, vec2i pos) {
     return a;
 }
 
+int animation_clone(animation *src, animation *dst) {
```
Comment (mrannanj) [inline src/resources/animation.c:56]: void function

**Review by mrannanj — CHANGES_REQUESTED** [review]

File: src/controller/net_controller.c [hunk 12]
```diff
@@ -101,50 +324,182 @@ int net_controller_tick(controller *ctrl, int ticks, ctrl_event **ev) {
...
+                                if(!data->synchronized) {
+                                    DEBUG("peer @ %d guessed our ticks INcorrectly! %d %d %d, actually  %d", peerticks,
+                                          start, peerguess, peerguess - start, ticks - start);
+                                }
+                                data->guesses = 0;
+                            }
+                            int newrtt = ticks - start;
+                            if(data->guesses >= 10 &&
+                               ((data->id == ROLE_SERVER && ticks % 37 == 0) ||
+                                (data->id == ROLE_CLIENT && ticks % 73 == 0)) &&
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/controller/net_controller.c:416]: they're just handy primes to avoid timing conflicts

File: src/game/protos/object.c (hunk 2 above)
Comment (Vagabond) [reply src/game/protos/object.c]: this is for the object's specializer clone call, not the generic object clone

File: src/game/game_player.c (hunk 3 above)
Comment (Vagabond) [reply src/game/game_player.c]: we null the HAR out when the fight ends, but we used to just store a pointer, now we store an object ID, so we need to check for the null pointer

File: src/game/scenes/arena.c (hunk 8 above)
Comment (Vagabond) [reply src/game/scenes/arena.c]: Removed this code.

File: src/game/scenes/arena.c [hunk 13]
```diff
@@ -795,20 +770,30 @@ void arena_free(scene *scene) {
...
+    }*/
     object_free(local->player1_portrait);
     omf_free(local->player1_portrait);
 
-    if(local->player2_portrait != NULL) {
+    /*if(local->player2_portrait != NULL) {
         sprite_free(local->player2_portrait->cur_sprite);
         omf_free(local->player2_portrait->cur_sprite);
-    }
+    }*/
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/game/scenes/arena.c:783]: I removed the need for them by properly tracking the sprite inside the object.

[the remark this answers is not in the document]
Comment (mrannanj) [reply src/game/scenes/arena.c:783]: Okay, so need remove the code instead of commenting out.

File: src/game/protos/scene.c (hunk 7 above)
Comment (Vagabond) [reply src/game/protos/scene.c:144]: I was debugging without asan, and it was hard to tell if I was freeing it

File: src/game/scenes/arena.c [hunk 14]
```diff
@@ -734,13 +709,13 @@ void arena_har_hook(har_event event, void *data) {
...
-    obj_har1 = game_player_get_har(game_state_get_player(scene->gs, 0));
-    obj_har2 = game_player_get_har(game_state_get_player(scene->gs, 1));
+    obj_har1 = game_state_find_object(scene->gs, game_player_get_har_obj_id(game_state_get_player(scene->gs, 0)));
+    obj_har2 = game_state_find_object(scene->gs, game_player_get_har_obj_id(game_state_get_player(scene->gs, 1)));
     har *har1, *har2;
     har1 = obj_har1->userdata;
     har2 = obj_har2->userdata;
 
-    if(scene->gs->role == ROLE_CLIENT) {
+    if(scene->gs->role == ROLE_CLIENT && false) {
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/game/scenes/arena.c:719]: I want to keep it, added a comment

**Review by mrannanj — APPROVED** [review]
